contracts/GSDINFT.sol

```diff
@@ -0,0 +1,200 @@
+  }
+
+  ///@notice Cancels the GSDI proposal. Sender must be borrower. GSDI must currently be in proposal. GSDI must be on the current chain. Burns the GSDI.
+  /// @param _id GSDI to cancel.
+  function cancel(uint256 _id) external mustBeInProposal(_id) override {
+    burnProposal(_id);
+    
+    emit Cancel(_id);
+  }
+
+  /// @notice Sends the GSDI to sender. The IGSDIWallet must be in proposal. GSDI must be on the current chain. Transfers price in currency from sender to borrower. Removes GSDI from proposal.
+  /// @notice If governance has enabled the 0.3% fee, the sender must also transfer the fee to the governance's treasury.
+  /// @dev Sender must approve the contract to transfer price of currency before call.
+  /// @param _id GSDI to purchase.
+  function purchase(uint256 _id) external mustBeInProposal(_id) override {
+    uint256 price = metadata[_id].price;
+    
+    IERC20 token = IERC20(metadata[_id].currency);
+    if (token.allowance(msg.sender, address(this)) < price) {
+      token.approve(address(this), price);
+    }
+
+    bytes memory transferPayload = abi.encodeWithSignature("transferFrom(address,address,unint256)", msg.sender, metadata[_id].borrower, price);
+    IGSDIWallet(metadata[_id].wallet).execute(metadata[_id].borrower, 1 ether, transferPayload);
+    metadata[_id].isInProposal = false;
+
+    emit Purchase(_id, msg.sender, price, metadata[_id].borrower);
+  }
+
+  /// @notice Sets the borrower for IGSDIWallet to executor. The IGSDIWallet must not be in proposal. GSDI must be on the current chain. Sender transfers face value in currency to current GSDI holder. Burns the GSDI.
+  /// @dev Sender must approve the contract to transfer face value of currency before call.
+  /// @param _id GSDI to cover.
+  function cover(uint256 _id) external override mustNotBeInProposal(_id) {
+    uint256 faceValue = metadata[_id].faceValue;
+    
+    metadata[_id].wallet.setExecutor(metadata[_id].borrower);
+
+    IERC20 token = IERC20(metadata[_id].currency);
+    if (token.allowance(msg.sender, address(this)) < faceValue) {
+      token.approve(address(this), faceValue);
+    }
+
+    bytes memory transferPayload = abi.encodeWithSignature("transferFrom(address,address,unint256)", msg.sender, metadata[_id].wallet.executor(), faceValue);
+    IGSDIWallet(metadata[_id].wallet).execute(metadata[_id].wallet.executor(), 1 ether, transferPayload);
+    burnProposal(_id);
+
+    emit Cover(_id, msg.sender, faceValue, metadata[_id].wallet.executor());
+  }
+
+  /// @notice Sets the lender for IGSDIWallet to executor. The GSDI must be held by sender. GSDI must be on the current chain. Must be after maturity. Burns the GSDI.
+  /// @param _id GSDI to seize.
+  function seize(uint256 _id) external override {
+    require(metadata[_id].wallet.executor() == msg.sender, "The GSDI must be held by sender.");
+    require(metadata[_id].maturity > block.timestamp, "Must be after maturity.");
+    burnProposal(_id);
```

Comment:
Before burn, the IGSDIWallet's executor must be set to the lender (msg.ender here)